distributed/distributed-schema.yaml

```diff
@@ -117,6 +117,27 @@ properties:
             description: |
               How frequently to balance worker loads
 
+          worker-saturation:
+            type: number
+            exclusiveMinimum: 0
+            description: |
+              Controls how many root tasks are sent to workers (like a `readahead`).
+
+              Up to worker-saturation * nthreads root tasks are sent to a
+              worker at a time. If `.inf`, all runnable tasks are immediately sent to workers.
+
+              Allowing oversaturation (> 1.0) means a worker will start running a new root task as
```

Comment:
```suggestion
              Allowing oversaturation (> 1.0) means a worker may start running a new root task as
```
The worker will execute the dependency task first if the other root task does not have enough time to reach the top of the ready queue.

Effectively, this means that, even with this set to .inf, as long as worker<->scheduler RTT is less than the execution time of a task, you should have nthreads+1 parallel streams per worker at any one time. 

In other words: this setting has exponentially diminishing returns, both in terms of throughput and memory usage. A higher setting will make more of a difference when task runtime is smaller than RTT. If task runtime is always larger than RTT, than 1.01 is all you need (*as long as you're rounding up in the code; didn't check yet*)

Corollary 1: using larger chunk sizes means longer computation times, which in turn means mitigated throughput degradation when you reduce this value.

Corollary 2: If you end up in the nthreads+1 best use case, using less workers with more threads each means less overall memory usage (as long as GIL contention isn't a problem). This is *in addition* to the savings caused by needing less data replications.

Corollary 3: GIL-heavy pure python workloads, which necessarily run on 1 thread per worker, would greatly benefit from setting this to 1 and make sure their task runtime is >> RTT.

You *may* want to point these things out. Although I'm conscious that there should be a line where this becomes too much of an infodump for less expert users.

TBH this feels more and more like an essay which should live in a dedicated sphinx page, with  a link here pointing to it.

Reply:
> The worker will execute the dependency task first if the other root task does not have enough time to reach the top of the ready queue

The other root task will very nearly always win:
1. As a root-ish task, it either has 0 dependencies, or <5 widely-shared dependencies that are almost certainly already in memory on the worker (since the previous task would have needed them as well). So there's no waiting for dependencies.
2. It's not just about worker<->scheduler RTT being less than the execution time of a task, it's also about the variance in execution time. For the first `worker.nthreads` root tasks (which all start at the same time), each task would also have to take >RTT seconds to complete longer than the _previous_ root task. Otherwise, the scheduler won't have had time to put a higher-priority task on the `ready` queue, so instead the next root task will run.

In practice, you'll have `nthreads * 2` parallel streams per worker at once _minimum_. (Anytime there's a network or scheduler delay in sending a new task, the worker will pop off another root task and run it, increasing that factor.) Here's a test demonstrating it on main:
```python
    @gen_cluster(nthreads=[("", 2)] * 4, client=True)
    async def test_graph_execution_width(c, s, *workers):
        class Refcount:
            "Track how many instances of this class exist; logs the count at creation and deletion"
    
            count: ClassVar[int] = 0
            lock: ClassVar[dask.utils.SerializableLock] = dask.utils.SerializableLock()
            log: ClassVar[list[int]] = []
    
            def __init__(self) -> None:
                sleep(0.1)  # significantly longer than scheduler<->worker round-trip time
                with self.lock:
                    type(self).count += 1
                    self.log.append(self.count)
    
            def __del__(self):
                with self.lock:
                    self.log.append(self.count)
                    type(self).count -= 1
    
        roots = [delayed(Refcount)() for _ in range(64)]
        passthrough = [delayed(slowidentity)(r) for r in roots]
        done = [delayed(lambda r: None)(r) for r in passthrough]
    
        fs = c.compute(done)
        await wait(fs)
        print(Refcount.log)
>       assert max(Refcount.log) <= sum(w.state.nthreads + 1 for w in workers)
E       AssertionError: assert 16 <= 12
E        +  where 16 = max([1, 2, 3, 4, 5, 6, ...])
E        +    where [1, 2, 3, 4, 5, 6, ...] = <class 'test_scheduler.test_graph_execution_width.<locals>.Refcount'>.log
E        +  and   12 = sum(<generator object test_graph_execution_width.<locals>.<genexpr> at 0x7f8610178510>)

distributed/tests/test_scheduler.py:348: AssertionError
------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 6, 7, 7, 7, 8, 9, 10, 10, 9, 9, 10, 10, 9, 8, 7, 6, 5, 5, 6, 6, 5, 5, 6, 6, 5, 5, 6, 7, 8, 9, 10, 11, 11, 11, 11, 10, 9, 8, 7, 7, 8, 9, 10, 11, 12, 12, 12, 12, 12, 12, 11, 10, 9, 8, 7, 6, 5, 4, 4, 5, 6, 7, 7, 7, 7, 6, 6, 7, 8, 8, 7, 6, 5, 5, 6, 6, 5, 5, 6, 6, 5, 5, 6, 7, 8, 9, 10, 10, 9, 9, 9, 9, 9, 8, 7, 6, 5, 4, 3]
```

And if I replace the `sleep(0.1)` with a GIL-holding `while time() < start + 0.1: pass` loop (to gum up the event loop on the scheduler), I get an execution width of 24 on my machine, up from 16.

---

But about the overall point:

> TBH this feels more and more like an essay which should live in a dedicated sphinx page, with a link here pointing to it.

Totally agreed. As we get more actual use of this, we'll want to write some dedicated documentation for it. I don't want to do too much of that without testing yet.

Though ideally, I don't actually see this as a long-lived parameter. I'm hoping we'll either find that some value (like 1.0, or 1.1) works well across the board, and remove the ability to tweak it, or that it'll stop being a global parameter, and instead more be something that's estimated (or annotated) per task.